### ISSUE TITLE
[typing] Add missing `__all__` to modules.

### ISSUE THREAD

**Issue by randolf-scholz** [pytorch/pytorch#131765]
### 🚀 The feature, motivation and pitch

`torch` ships with the `py.typed` marker, yet many modules lack a `__all__` declaration. 
This leads to problems with type checkers:

```python
from torch import optim
opt_type = optim.Adam  # pyright: error: "Adam" is not exported from module "torch.optim"
```

According to the [typing spec](https://typing.readthedocs.io/en/latest/guides/libraries.html#library-interface-public-and-private-symbols):

> Imported symbols are considered private by default
> A module can expose an __all__ symbol at the module level that provides a list of names that are considered part of the interface. 

Therefore, modules that reexport symbols such as https://github.com/pytorch/pytorch/blob/main/torch/optim/__init__.py need to declare `__all__`.

**Comment by malfet** [comment]
Thank you for reporting. If you already have a fix, please do not hesitate to propose a PR

**Comment by albanD** [comment]
Note that our current public API definitely doesn't require `__all__` to be defined https://github.com/pytorch/pytorch/wiki/Public-API-definition-and-documentation

**Comment by randolf-scholz** [comment]
@albanD Well, as I said, the problem with this is it violates assumptions made by 3rd party tools such as `pyright`, based on the official typing specification.

**Comment by randolf-scholz** [comment]
@malfet I made a fix #131800. It's rather large as I updated over 150 `__all__` definitions.

**Comment by albanD** [comment]
I understand that `pyright` does that but the typing specifications you shared above don't actually make that claim?
Is `pyright` more strict than the general CPython typing module rules?
If so, do these properly align with other typecheckers? We don't use pyright AFAIK and use mypy/black/flake8/ufmt. Will this conflict with them?

**Comment by ezyang** [comment]
@albanD pyright is very popular downstream so we should support them anyway even if we don't use it internally

**Comment by albanD** [comment]
@randolf-scholz can you share how you get this error btw, installing pyright and running it on the code you shared gives me `0 errors, 0 warnings, 0 informations`.

**Comment by randolf-scholz** [comment]
@albanD Not sure why it doesn't show up for you; I created an example repro here: https://github.com/randolf-scholz/torch_131765

**Comment by randolf-scholz** [comment]
I simplified my main PR and split up some smaller edits into separate PRs:

- [ ] https://github.com/pytorch/pytorch/pull/131800
- [ ] https://github.com/pytorch/pytorch/pull/132403
- [ ] https://github.com/pytorch/pytorch/pull/132404
- [ ] https://github.com/pytorch/pytorch/pull/132405

**Comment by albanD** [comment]
Ho I guess because we already added it in https://github.com/pytorch/pytorch/blob/6c1f1563e1b2fa54673033a5aed8fabd9fb28bc2/torch/optim/__init__.py#L44 ?

Also I can't seem to find the exact definition of what pyright considers public or not. Do you know where that is by any chance? 
The typing spec you shard above aligns with what we do by saying that `__all__` is optional and, when not provided, the other rules apply (they define these extra rules based on static code of import while we define them based on runtime object __module__ attribute).

It would be great to get the exact specs of what we need to comply with so we can properly update our rules and lint to make sure we continue following them!

**Comment by randolf-scholz** [comment]
@albanD Yes, it seems like @ringohoffman already patched `optim/__init__.py` in https://github.com/pytorch/pytorch/pull/131959 last week.

`pyright` follows the spec here. However, reading it again, it seems like the wording could be improved.
What the spec says is that if there is no `__all__`, then star-imports and imports that are redundantly renamed are considered public. (unless they start with underscore)

On the other hand, the torch spec says that all objects with a `__module__` attribute that starts with `torch.`  are considered public.

In my example repo, if one replaces

```python
from .adam import Adam
```

with 

```python
from .adam import Adam as Adam
```

then `pyright` considers `Adam` public.

**Comment by randolf-scholz** [comment]
pylance also adds some visual highlighting clue in this case:


![image](https://github.com/user-attachments/assets/3554383e-cd28-4dd4-a7ec-89a74458fdb4)

**Comment by albanD** [comment]
Thanks for the details!

Yes the any module that starts with `torch.` is to make the transition easier as there are many places where we do re-export in parent namespaces.
Also this helps appease sphinx doc coverage (that we use) where it will consider an API to be public if its `__module__` is the current module.
If we do `from .adam import Adam as Adam`, does that mean that sphinx will see two public APIs (that must be documented): torch.optim.Adam and torch.optim.adam.Adam ?

**Comment by randolf-scholz** [comment]
I'm not sure, but it seems that was already changed in https://github.com/pytorch/pytorch/pull/127703 anyway.

**Comment by dcolascione** [comment]
`__all__` also doesn't include the dtypes that `_C` magically adds to the `torch` module.

**Comment by adamtheturtle** [comment]
One more concrete case for this issue: the top-level `torch` module's `__all__` in [`torch/__init__.py`](https://github.com/pytorch/pytorch/blob/main/torch/__init__.py) lists `Tensor` (the class) but not `tensor` (the factory) or `from_numpy` — both are injected from `torch._C` at runtime. pyright 1.1.409 flags these where 1.1.408 did not.

**Minimal repro** (`pip install torch pyright==1.1.409`):

```python
import torch

t = torch.tensor(data=[1.0, 2.0, 3.0])
u = torch.from_numpy
```

**pyright 1.1.408:** clean.
**pyright 1.1.409:**

```
error: "tensor" is not exported from module "torch" (reportPrivateImportUsage)
error: "from_numpy" is not exported from module "torch" (reportPrivateImportUsage)
```
